okhttp/src/main/java/okhttp3/internal/Transmitter.java

```diff
@@ -113,283 +114,161 @@ private Address createAddress(HttpUrl url) {
-      eventListener.responseHeadersStart(call);
-      return streamAllocation.codec().readResponseHeaders(expectContinue);
-    } catch (IOException e) {
-      streamAllocation.streamFailed(e);
-      throw e;
+  /**
+   * Remove the transmitter from the connection's list of allocations. Returns a socket that the
+   * caller should close.
+   */
+  public @Nullable Socket releaseConnectionNoEvents() {
+    assert (Thread.holdsLock(connectionPool));
+
+    int index = -1;
+    for (int i = 0, size = this.connection.transmitters.size(); i < size; i++) {
+      Reference<Transmitter> reference = this.connection.transmitters.get(i);
+      if (reference.get() == this) {
+        index = i;
+        break;
+      }
     }
-  }
-
-  public void responseHeadersEnd(Response response) {
-    eventListener.responseHeadersEnd(call, response);
-  }
 
-  public ResponseBody openResponseBody(Response response) throws IOException {
-    try {
-      streamAllocation.eventListener.responseBodyStart(streamAllocation.call);
-      String contentType = response.header("Content-Type");
-      HttpCodec codec = streamAllocation.codec();
-      long contentLength = codec.reportedContentLength(response);
-      Source rawSource = codec.openResponseBodySource(response);
-      ResponseBodySource source = new ResponseBodySource(rawSource, contentLength);
-      return new RealResponseBody(contentType, contentLength, Okio.buffer(source));
-    } catch (IOException e) {
-      streamAllocation.streamFailed(e);
-      throw e;
-    }
-  }
+    if (index == -1) throw new IllegalStateException();
 
-  public DeferredTrailers deferredTrailers() {
-    return new DeferredTrailers() {
-      HttpCodec codec = streamAllocation.codec();
+    RealConnection released = this.connection;
+    released.transmitters.remove(index);
+    this.connection = null;
 
-      @Override public Headers trailers() throws IOException {
-        return codec.trailers();
+    if (released.transmitters.isEmpty()) {
+      released.idleAtNanos = System.nanoTime();
+      if (connectionPool.connectionBecameIdle(released)) {
+        return released.socket();
       }
-    };
-  }
-
-  public boolean supportsUrl(HttpUrl url) {
-    return streamAllocation.connection().supportsUrl(url);
-  }
+    }
 
-  public void acquireConnection(RealConnection connection) {
-    streamAllocation.transmitterAcquireConnection(connection);
+    return null;
   }
 
-  public RealConnection connection() {
-    return streamAllocation.connection();
+  public void exchangeDoneDueToException() {
+    synchronized (connectionPool) {
+      if (noMoreExchanges) throw new IllegalStateException();
+      exchange = null;
+    }
   }
 
-  public void responseBodyComplete(long bytesRead, IOException e) {
-    if (e != null) {
-      streamAllocation.streamFailed(e);
-    }
-    if (streamAllocation != null) {
-      streamAllocation.responseBodyComplete(bytesRead, e);
+  public void exchangeDone(@Nullable IOException e) {
```

Comment:
FYI @oldergod this is the method we didn’t have before that we really needed. With duplex we’ll be able to have the request stream end after the response stream ends.